src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs

```diff
@@ -358,9 +358,10 @@ private ClusterDescription ProcessReplicaSetChange(ClusterDescription clusterDes
             clusterDescription = EnsureServers(clusterDescription, args.NewServerDescription, newServers);
 
             if (args.NewServerDescription.CanonicalEndPoint != null &&
-                !EndPointHelper.Equals(args.NewServerDescription.CanonicalEndPoint, args.NewServerDescription.EndPoint))
+                !EndPointHelper.Equals(args.NewServerDescription.CanonicalEndPoint, args.NewServerDescription.EndPoint) &&
+                args.NewServerDescription.Type != ServerType.ReplicaSetPrimary)
```

Comment:
A useful [note from the spec on updateRsFromPrimary](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#updatersfromprimary):
>A note on checking "me": Unlike updateRSWithPrimaryFromMember, there is no need to remove the server if the address is not equal to "me": since the server address will not be a member of either "hosts", "passives", or "arbiters", the server will already have been removed.

Reply:
Although it might be more accurate to say "the server will be removed" as part of 

> for each server in topologyDescription.servers:
    if server.address not in description's "hosts", "passives", or "arbiters":
        remove the server and stop monitoring it